sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -561,6 +561,12 @@ object ColumnPruning extends Rule[LogicalPlan] {
     case d @ DeserializeToObject(_, _, child) if !child.outputSet.subsetOf(d.references) =>
       d.copy(child = prunedChild(child, d.references))
 
+    case p @ Project(_, s: SerializeFromObject) if p.references != s.outputSet =>
+      val prunedSerializer = s.serializer.filter { expr =>
+        p.references.contains(expr.toAttribute)
```

Comment:
why? we are not going to get intersected attributes. we want to get part of serializers which are referred by top projection.

Reply:
I'd just want to avoid regenerating the `p.references` `AttributeSet` for every expr, which may become expensive if there are a lot of exprs. If intersect doesn't work, maybe just a:
```
val prunedRefs = p.references
val prunedSerializer = s.serializer.filter(prunedRefs.contains)
```

Reply:
ok. I see. thanks.